source/common/network/connection_balancer_impl.cc

```diff
@@ -0,0 +1,40 @@
+#include "common/network/connection_balancer_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+void ExactConnectionBalancerImpl::registerHandler(BalancedConnectionHandler& handler) {
+  absl::MutexLock lock(&lock_);
+  handlers_.push_back(&handler);
+}
+
+void ExactConnectionBalancerImpl::unregisterHandler(BalancedConnectionHandler& handler) {
+  absl::MutexLock lock(&lock_);
+  handlers_.erase(std::find(handlers_.begin(), handlers_.end(), &handler));
+}
+
+ConnectionBalancer::BalanceConnectionResult
+ExactConnectionBalancerImpl::balanceConnection(ConnectionSocketPtr&& socket,
+                                               BalancedConnectionHandler& current_handler) {
+  BalancedConnectionHandler* min_connection_handler = nullptr;
+  {
+    absl::MutexLock lock(&lock_);
+    for (const auto handler : handlers_) {
+      if (min_connection_handler == nullptr ||
+          handler->numConnections() < min_connection_handler->numConnections()) {
+        min_connection_handler = handler;
+      }
+    }
+
+    min_connection_handler->incNumConnections();
+  }
+
+  if (min_connection_handler != &current_handler) {
+    min_connection_handler->post(std::move(socket));
+    return ConnectionBalancer::BalanceConnectionResult::Rebalanced;
+  }
```

Comment:
This piece of code seems possible to be common code cross all types of balancer. 
Reading the comment on ConnectionHandler::incNumConnections(), there seems to be a constrain about when this interface should be called by balancer. Can we modify the balancer interface to have the constrain built into it and just expose a method like pickTargetHandler() for different implementations.

Reply:
Good call will update.

Reply:
Sorry, probably I didn't make it clear. I was suggesting to make ConnectionBalancer::balanceConnection() a defined member function which calls pickTargetHandler() whose implementation is upto different balancer impl. Because I could imagine all the balancers' balanceConnection() will need to do connection accounting and post(connection_socket)  if the returned targed_handler is different from current one. If so, why not just put this logic into ConnectionHandler interface directly.

Reply:
I believe I did this in the code I pushed?

Reply:
It's still a bit different. You replaced balanceConnection() with pickTargetHandler() and made onAcceptWorker() to call the latter, and then post(). In the two existing balancer impl's, pickTargetHandler() still need to increment #connections.  

What I have in mind is something like below:
```
class ConnectionBalancer {
public:
   BalanceResult balanceConnection(current_handler) {
     BalancedConnectionHandler& target_handler = pickTargetHandler(current_handler);
     if (target_handler != current_handler) {
        // Hand over this connection
        target_handler.incNumConnections();
        target_handler.post[](){...});
        return Rebalanced;
     } 
     return Continue;
   }

  // Differs across different impl's.
  BalancedConnectionHandler& pickTargetHandler(BalancedConnectionHandler&) PURE;

  // Other interfaces.
};
```

Reply:
See the comments I added in connection_balancer.h for incNumConnections() and pickTargetHandler(). Unfortunately, I don't think there is a good way to move the incNumConnections() call into a common place, as the locking differs between balancers. In the exact balancer the lock must be held during incrementing or there will not be exact balancing.

My preference is to land this and we can simplify/iterate in later changes if it becomes possible?